### PR TITLE
Add rate and channels to ALH config

### PR DESCRIPTION
**Pull request by bardliao**
FW needs to know rate and channels for ALH dai. @slawblauciak will create the corresponding PR on FW side.

### CODE REVIEW

**Comment by slawblauciak** [comment]
I will push the FW change next monday. Our CI is still being in the process of clean up.

**Comment by plbossart** [comment]
Can someone please explain why this is needed for ALH, and why it is needed for the HDaudio link DMA? we've worked for other 6 months without the information, why is this needed and why now?

**Comment by kv2019i** [comment]
@singalsu wrote:
> Can someone please explain why this is needed for ALH, and why it is needed for the HDaudio link DMA? we've worked for other 6 months without the information, why is this needed and why now?

@singalsu can explain better, but I've understood these parameters are used by some audio processing elements in the pipeline, e.g. ASRC. So if you don't have those in the pipeline, things have worked ok. But e.g. AHL + ASRC pipeline will hit this.

**Comment by plbossart** [comment]
> @singalsu wrote:
> 
> > Can someone please explain why this is needed for ALH, and why it is needed for the HDaudio link DMA? we've worked for other 6 months without the information, why is this needed and why now?
> 
> @singalsu can explain better, but I've understood these parameters are used by some audio processing elements in the pipeline, e.g. ASRC. So if you don't have those in the pipeline, things have worked ok. But e.g. AHL + ASRC pipeline will hit this.

We are just trying to make ALH work for now, ASRC is not needed in the foreseable future on such platforms, all clocks are owned by the Intel chip. same for HDAudio, the clock is owned by the controller so why worry about ASRC?

**Comment by slawblauciak** [comment]
@plbossart demux also suffers from this. Not just ASRC. And we need demux for synchronized playback.

**Comment by singalsu** [comment]
@plbossart Also normal SRC needs the parameters from downstream. DAI parameters are propagated to pipeline from downstream.

File: sound/soc/sof/intel/hda.c [hunk 1]
```diff
@@ -68,6 +68,8 @@ static int sdw_params_stream(struct device *dev,
 	config.type = SOF_DAI_INTEL_ALH;
 	config.dai_index = (link_id << 8) | (d->id);
 	config.alh.stream_id = alh_stream_id;
+	config.alh.rate = params_rate(params_data->hw_params);
+	config.alh.channels = params_channels(params_data->hw_params);
```
Comment (plbossart) [inline sound/soc/sof/intel/hda.c]: @bardliao who configures the rate and channels present in the hw_params? it's not part of the topology definitions, so where does this come from? the front-end?
Say e.g. that you want to push 1ch on ALH1.2 and 1ch on ALH 2.2, how would this work?

I am not against providing the info to the firmware, but it seems a bit too magical for now. Thanks for details.

Comment (bardliao) [reply sound/soc/sof/intel/hda.c]: > @bardliao who configures the rate and channels present in the hw_params? it's not part of the topology definitions, so where does this come from? the front-end?
> Say e.g. that you want to push 1ch on ALH1.2 and 1ch on ALH 2.2, how would this work?

It comes from soc core. The flow is soc_pcm_hw_params() -> soc_soc_dai_hw_params() ->intel_hw_params() -> intel_params_stream() -> sdw_params_stream(). 
All ALH dais on the same PCM will get the same hw_params. But, we can configure it to what we want in sdw_params_stream().

**Comment by slawblauciak** [comment]
FW change wasn't up yet. It is now:
https://github.com/thesofproject/sof/pull/2666

File: sound/soc/sof/intel/hda.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda.c]: @bardliao aren't you making big assumptions about what the demux does here and how it's configured? The SOC code  does not have any notion of demux/aggregration, so if the demux configuration was to pass 1 channel to ALH1.2 and 1 channel to ALH2.2, then what would the firmware do?

We also need to consider the capture case, if you have multiple devices that capture from 2 or more devices, then do you really want to send to each ALH DAI the configuration that comes from the SOC core?

I must be missing something here.

File: sound/soc/sof/topology.c [hunk 2]
```diff
@@ -3115,13 +3125,26 @@ static int sof_link_alh_load(struct snd_soc_component *scomp, int index,
...
+			le32_to_cpu(private->size));
+		return ret;
+	}
+
 	/* init IPC */
 	config->hdr.size = size;
 
+	dev_dbg(scomp->dev, "ALH config rate %d channels %d\n",
+		config->alh.rate, config->alh.channels);
+
```
Comment (plbossart) [inline sound/soc/sof/topology.c:3127]: I think this is on the right path, but then the IPC that's done in the hw_params stage would not know about this configuration?

File: sound/soc/sof/intel/hda.c (hunk 1 above)
Comment (bardliao) [reply sound/soc/sof/intel/hda.c]: > @bardliao aren't you making big assumptions about what the demux does here and how it's configured? The SOC code does not have any notion of demux/aggregration, so if the demux configuration was to pass 1 channel to ALH1.2 and 1 channel to ALH2.2, then what would the firmware do?
> 
> We also need to consider the capture case, if you have multiple devices that capture from 2 or more devices, then do you really want to send to each ALH DAI the configuration that comes from the SOC core?
> 
> I must be missing something here.

Let's config it on topology :)

**Comment by bardliao** [comment]
The corresponding topology PR is https://github.com/thesofproject/sof/pull/2675

File: sound/soc/sof/topology.c (hunk 2 above)
Comment (bardliao) [reply sound/soc/sof/topology.c:3127]: > I think this is on the right path, but then the IPC that's done in the hw_params stage would not know about this configuration?

I discussed it with @slawblauciak and we will ignore the configuration if both rate and channels are set to 0 in the configuration .

**Comment by bardliao** [comment]
> we need an ABI MINOR change as well. @lgirdwood FYI

I will create a similar PR for HDA, and  I think it is worth to use the same ABI change.

**Comment by bardliao** [comment]
> > we need an ABI MINOR change as well. @lgirdwood FYI
> 
> I will create a similar PR for HDA, and I think it is worth to use the same ABI change.

Ah, there is no reserved field in [struct sof_ipc_dai_hda_params](https://github.com/thesofproject/linux/blob/topic/sof-dev/include/sound/sof/dai-intel.h#L85). So I guess it will be an ABI MAJOR change, if we want to add any param there, right? BTW, @singalsu @slawblauciak what params will we need?

**Comment by plbossart** [comment]
> > > we need an ABI MINOR change as well. @lgirdwood FYI
> > 
> > 
> > I will create a similar PR for HDA, and I think it is worth to use the same ABI change.
> 
> Ah, there is no reserved field in [struct sof_ipc_dai_hda_params](https://github.com/thesofproject/linux/blob/topic/sof-dev/include/sound/sof/dai-intel.h#L85). So I guess it will be an ABI MAJOR change, if we want to add any param there, right? BTW, @singalsu @slawblauciak what params will we need?

or the firmware can allocate a bigger size and deal with backwards compatibility internally, as done for another PR recently

**Comment by kv2019i** [comment]
> or the firmware can allocate a bigger size and deal with backwards compatibility internally, as done for another PR recently

Yes, @bardliao you can add parameters at end of IPC structs with a minor ABI version bump.

**Comment by slawblauciak** [comment]
@bardliao definitely at least channels. Maybe rate too. Dunno about bit depth, I'm not familiar with HDA enough to be able to tell.

**Comment by kv2019i** [comment]
@slawblauciak wrote:
> @bardliao definitely at least channels. Maybe rate too. Dunno about bit depth, I'm not familiar with HDA enough to be able to tell.

Ok that raises a concern. For @singalsu as well. If the FW arch assumes that DAIs can tell the channel count and rate to pipeline components to work correctly, wouldn't these parameters belong to sof_ipc_dai_config? I mean if we now fix ALH and HDA to have this data, how about IMX DAIs (@dbaluta )?

**Comment by bardliao** [comment]
I added the HDA change to this PR with ABI change. @lgirdwood Could you confirm the ABI number?

File: include/uapi/sound/sof/abi.h [hunk 3]
```diff
@@ -26,7 +26,7 @@
 
 /* SOF ABI version major, minor and patch numbers */
 #define SOF_ABI_MAJOR 3
-#define SOF_ABI_MINOR 15
+#define SOF_ABI_MINOR 16
```
Comment (slawblauciak) [inline include/uapi/sound/sof/abi.h]: The FW change was made with minor version 15

Comment (bardliao) [reply include/uapi/sound/sof/abi.h]: > The FW change was made with minor version 15

Isn't minor version 15 for "ipc: cc_version: use fixed length for CC_DESC"? And I think the HDA change is not ready on FW side, right? I am not sure if we can use the same minor version with more than one ipc changes.

Comment (slawblauciak) [reply include/uapi/sound/sof/abi.h]: The FW change for ALH has already been merged and it uses minor version 15. Maybe better to split this change in that case.

Comment (lgirdwood) [reply include/uapi/sound/sof/abi.h]: This can still go in 15. Please update to use 15.

Comment (bardliao) [reply include/uapi/sound/sof/abi.h]: > This can still go in 15. Please update to use 15.

Got it. Thanks @lgirdwood

**Comment by bardliao** [comment]
Updated @plbossart @kv2019i @ranj063 @slawblauciak @singalsu

**Comment by bardliao** [comment]
HDA part removed. @plbossart @kv2019i @ranj063 I think the PR is ready for merge now. @lgirdwood has confirmed that this can still go in ABI 3.15.0. I will create another PR for the HDA part.

**Comment by kv2019i** [comment]
For some reason the FW change does not show on the ABI classifier, but it's indeed merged: https://github.com/thesofproject/sof/pull/2666 
So looks ok, let's merge this.

**Comment by kv2019i** [comment]
Note to @bardliao -- I didn't notice this before, but the git commit messages are really, really short. I think we should have a summary of why the change is needed, on the Linux side commits as well.  Now we have good description of the problem and solution in FW side commit messages, but on Linux side we just have a "FW needs these params" with no reference to what problem is being solved.

As I didn't catch this before and other reviewer's seem ok, I'll go ahead and merge. But let's pay more attention this ins future patches.

**Pull request merged**